_drafts/2023-08-10-functional-error-handling-in-kotlin-part-3.md

```diff
@@ -160,6 +160,17 @@ In fact, the compilation error is:
 Type mismatch: The inferred type is GenericError, but JobNotFound was expected
 ```
 
+At first, it could seem a limitation and a little misleading. However, it's not. In fact, the `Raise<E>` type is a **context**, giving us some capabilities once we run a function in its scope. We should remember that in Kotlin, we can run a lambda inside a context using the `with` function:
+
+```kotlin
+val raiseScope = TODO()
+with(raiseScope) {
+    val job = appleJob()
+}
+```
+
+In the above code, we can run the `appleJob` function inside the `Raise<E>` context. In this sense, the `Raise<E>` type is similar to the `CoroutineScope` type. The scope lets us access features we can't access outside the scope. In the case of the `CoroutineScope`, we can run suspending functions using structural concurrency. In the case of the `Raise<E>` scope, we can run functions that can raise errors of type `E`.
+
```

Comment:
I am still on the fence on this step of the tutorial. Your new text improves the explanation, but I'm wondering if we are overcomplicating things from the beginning.

Right now, my outstanding questions as a reader are 1) "how would I build the raise scope myself?", and 2) "why this particular structure?"

For 1), creating the raise scope yourself is only done by Arrow with that DefaultRaise. Which means that this structure is uncallable from "user space" at the moment. We're introducing unnecessary complexity and simplify it later, which strikes me as a bit artificial.

For 2), I cannot make up a compelling argument for this complex structure this early, as opposed to the other forms of "errors" that we've discussed in the other articles.

The only way I can justify the complexity to the reader is a "don't worry, trust me". Something of the form:

> Right now, this Raise context might seem a little complex, artificial and uncallable. How would we build a Raise context to call this function? Why this convoluted structure where we have other forms of error handling (links to the other articles)? So many questions! Walk with me and things will become clear. _Raise is a powerful structure that will allow us to handle typed errors without the boilerplate of "catching" them or storing them in explicit data structures._ Keep that in mind as we walk through this tutorial, it'll be worth it.

Reply:
Ok, I’ll try to find a solution during the weekend.

Thanks for the feedback.